web-api/src/mind.rs

```diff
@@ -231,6 +287,121 @@ where
+    let mut rng = thread_rng();
+    // define how many documents to sample
+    let amount_of_doc_used_to_prepare = 1;
+
+    // define how many documents to rerank
+    let amount_of_doc_used_to_rerank = 30;
+
+    //probability of a document being clicked if it is interesting to the user
+    let click_probability = 0.5;
+
+    //define how many iterations to perform
+    let iterations = 10;
+
+    // define how many documents to use to calculate NDCG
+    let nranks = vec![3, 5];
+
+    // create a 3d array to store the results
+    let mut results = Array::zeros((nranks.len(), iterations, 0));
+
+    // read in users' interests from json file
+    let users_interests = UsersInterests::new("user_categories.json").unwrap();
+    // read in documents from tsv file
+    let document_provider = DocumentProvider::new("news.tsv").unwrap();
+    // iterate over users' interests and select documents that are relevant to the user
+    for user_interest in users_interests.user_interests {
+        // get all documents that match user's interests
+        let interesting_documents = document_provider.get_all_interest(&user_interest);
+        // sample documents that will be used to prepare the reranker from interesting documents
+        let documents_to_prepare =
+            interesting_documents.choose_multiple(&mut rng, amount_of_doc_used_to_prepare);
+
+        let mut ndcgs = Array::zeros((nranks.len(), 0));
+
+        // Placeholder for interacting with the sampled news (preparing the reranker)
+        for document in documents_to_prepare {
+            println!("The document {:?} was interacted with.", document);
+        }
+
+        // running iterations
+        for _ in 0..iterations {
+            // Sample random documents
+            let sampled_documents = document_provider.sample(amount_of_doc_used_to_rerank);
+            //Placeholder for reranking the sampled documents
+            for document in &sampled_documents {
+                println!("The document {:?} was reranked.", document);
+            }
+            // Assing labels to the documents
+            let labels = score_documents(&sampled_documents, &user_interest);
+            // Calculate NDCG
+            let iteration_ndcg = ndcg(&labels, &nranks);
+
+            ndcgs
+                .push(Axis(1), ArrayView::from(&iteration_ndcg))
+                .unwrap();
+
+            //placeholder for interacting with the sampled news
+            for document in sampled_documents {
+                if document.is_interesting(&user_interest) {
+                    // user clicked on the document if the probability is greater than click_probability
+                    if rng.gen::<f32>() < click_probability {
+                        println!("The document {:?} was clicked.", document);
+                    }
+                }
+            }
+        }
+        // add ndcgs for the user to the results
+        results.push(Axis(2), ArrayView::from(&ndcgs)).unwrap();
```

Comment:
I have tried to implement this, but I will leave it unresolved as I am not sure if I have done this correctly according to the instructions

Reply:
it's fine the way you implemented it now 👍  the other suggestions are more into the direction of avoiding to allocate the `results` array at all. this optimization is not necessary for our test cases, it was more a suggestion in case you want to look further into `Write`rs and the corresponding traits.